File: workplace2_src/clipTest.java
import example.clip;
import org.junit.Test;

import static org.junit.Assert.assertTrue;

public class clipTest {
    private void applyMR_Assert(Integer[] originalInput, int[] originalResult) {

        //MR1:数组元置换（打乱顺序）
        Integer[] transformInput1 = MetamorphicTestGenerator1.applyMR1(originalInput);
        int[] transformResult1 = clip.clip_m(transformInput1, 1, 5);
        //MR2:数组元素常数加法
        Integer[] transformInput2 = MetamorphicTestGenerator1.applyMR2(originalInput);
        int[] transformResult2 = clip.clip_m(transformInput2, 1, 5);
        //MR3_1:加入单位元不变性（加法的单位元0）
        Integer[] transformInput3_1 = MetamorphicTestGenerator1.applyMR3_1(originalInput);
        int[] transformResult3_1 = clip.clip_m(transformInput3_1, 1, 5);
        //MR3_2:加入单位元不变性（乘法的单位元1）
        Integer[] transformInput3_2 = MetamorphicTestGenerator1.applyMR3_2(originalInput);
        int[] transformResult3_2 = clip.clip_m(transformInput3_2, 1, 5);
        //MR4:数组元素取倒数
//        Double[] transformInput4 = MetamorphicTestGenerator1.applyMR4(originalInput);
//        Integer[] transformInput4_1 = new Integer[transformInput4.length];
//        for (int i = 0; i < transformInput4.length; i++) {
//            transformInput4_1[i] = transformInput4[i].intValue();
//        }
//        int[] transformResult4 = clip.clip_m(transformInput4_1);
        //MR5:数组缩放变换
        Integer[] transformInput5 = MetamorphicTestGenerator1.applyMR5(originalInput, 2);
        int[] transformResult5 = clip.clip_m(transformInput5, 1, 5);
        //MR6:数组反转变换
        Integer[] transformInput6 = MetamorphicTestGenerator1.applyMR6(originalInput);
        int[] transformResult6 = clip.clip_m(transformInput6, 1, 5);
        //MR7_1:中立操作的恒等变换（所有元素乘以1）
        Integer[] transformInput7_1 = MetamorphicTestGenerator1.applyMR7_1(originalInput);
        int[] transformResult7_1 = clip.clip_m(transformInput7_1, 1, 5);
        //MR7_2:中立操作的恒等变换（所有元素加上0）
        Integer[] transformInput7_2 = MetamorphicTestGenerator1.applyMR7_2(originalInput);
        int[] transformResult7_2 = clip.clip_m(transformInput7_2, 1, 5);
        //MR8:重复输入数组
        Integer[] transformInput8 = MetamorphicTestGenerator1.applyMR8(originalInput);
        int[] transformResult8 = clip.clip_m(transformInput8, 1, 5);
        //MR9:复合转换一致性
        Integer[] transformInput9 = MetamorphicTestGenerator1.applyMR9(originalInput, 3);
        int[] transformResult9 = clip.clip_m(transformInput9, 1, 5);
        //MR10:单调性检验
        Integer[] transformInput10 = MetamorphicTestGenerator1.applyMR10(originalInput);
        int[] transformResult10 = clip.clip_m(transformInput10, 1, 5);
        //MR11:边界值替换(把最大值替换成0)
        Integer[] transformInput11 = MetamorphicTestGenerator1.applyMR11(originalInput);
        int[] transformResult11 = clip.clip_m(transformInput11, 1, 5);
        //MR12:数值取反变换
//        Integer[] transformInput12 = MetamorphicTestGenerator1.applyMR12(originalInput);
//        int[] transformResult12 = clip.clip_m(transformInput12);
        //MR13:微小增量调整
//        Double[] transformInput13 = MetamorphicTestGenerator1.applyMR13(originalInput);
//        Integer[] transformInput13_1 = new Integer[transformInput13.length];
//        for (int i = 0; i < transformInput13_1.length; i++) {
//            transformInput13_1[i] = transformInput13[i].intValue();
//        }
//        double transformResult13 = clip.clip_m(transformInput13_1);
        //MR14:移除元素的效果（移除最大值）
        Integer[] transformInput14 = MetamorphicTestGenerator1.applyMR14(originalInput);
        int[] transformResult14 = clip.clip_m(transformInput14, 1, 5);
        //MR15:类三角函数的周期性
//        Integer[] transformInput15 = MetamorphicTestGenerator1.applyMR15(originalInput);
//        int[] transformResult15 = clip.clip_m(transformInput15);
        //MR16:重复值稳健性(复制输入中的一个元素)
        Integer[] transformInput16 = MetamorphicTestGenerator1.applyMR16(originalInput);
        int[] transformResult16 = clip.clip_m(transformInput16, 1, 5);
        //MR19:输入重复（元素复制）将元素a重复多次插入序列中
        Integer[] transformInput19 = MetamorphicTestGenerator1.applyMR19(originalInput, 2);
        int[] transformResult19 = clip.clip_m(transformInput19, 1, 5);
        //MR20:边界值灵敏度（给最小值增加一个极小值）
        Double[] transformInput20 = MetamorphicTestGenerator1.applyMR20(originalInput);
        Integer[] transformInput20_1 = new Integer[transformInput20.length];
        for (int i = 0; i < transformInput20_1.length; i++) {
            transformInput20_1[i] = transformInput20[i].intValue();
        }
        int[] transformResult20 = clip.clip_m(transformInput20_1, 1, 5);
        //MR22:应用恒等变换
        Integer[] transformInput22 = MetamorphicTestGenerator1.applyMR22(originalInput);
        int[] transformResult22 = clip.clip_m(transformInput22, 1, 5);
        //----------------------------------------------------------
//        //OR1:和应该保持不变
        assertTrue(sum(originalResult) == sum(transformResult1));
        //OR2:源输出和小于等于后续输入
        assertTrue(sum(originalResult) <= sum(transformResult2));
        //OR3_1:和应该增大
        assertTrue(sum(originalResult) <= sum(transformResult3_1));
        //OR3_2:和应该增大
        assertTrue(sum(originalResult) <= sum(transformResult3_2));
        //OR4:和应该减少或保持不变
//        assertTrue(originalResult >= transformResult4);
        //OR5:源输出小于等于后续输出
        assertTrue(sum(originalResult) <= sum(transformResult5));
        //OR6:源输出和等于后续输出和
        assertTrue(sum(originalResult) == sum(transformResult6));
        //OR7_1:源输出和等于后续输出和
        assertTrue(sum(originalResult) == sum(transformResult7_1));
        //OR7_2:源输出和等于后续输出和
        assertTrue(sum(originalResult) == sum(transformResult7_2));
        //OR8:源输出小于等于后续输出
        assertTrue(sum(originalResult) <= sum(transformResult8));
        //OR9:源输出小于等于后续输出
        assertTrue(sum(originalResult) <= sum(transformResult9));
        //OR10:源输出小于等于后续输出
        assertTrue(sum(originalResult) <= sum(transformResult10));
        //OR11:源输出大于等于后续输出
        assertTrue(sum(originalResult) >= sum(transformResult11));
        //OR12:源输出负数等于后续输出
//        assertTrue(-originalResult == transformResult12);
//        OR13:源输出小于等于等于后续输出
//        assertTrue(originalResult <= transformResult13);
        //OR14:源输出大于等于等于后续输出
        assertTrue(sum(originalResult) >= sum(transformResult14));
        //OR15:源输出大于等于等于后续输出
//        assertTrue(originalResult >= transformResult15);
        //OR16:源输出小于等于等于后续输出
        assertTrue(sum(originalResult) <= sum(transformResult16));
        //OR19:源输出小于等于等于后续输出
//        System.out.println("转换前的数组应该是：" + Arrays.toString(originalResult));
//        System.out.println("转换后的数组应该是：" + Arrays.toString(transformResult19));
        assertTrue(sum(originalResult) <= sum(transformResult19));
        //OR20:源输出小于等于后续输出
//        System.out.println("转换前的数组应该是：" + Arrays.toString(originalResult));
//        System.out.println("转换后的数组应该是：" + Arrays.toString(transformResult20));
        assertTrue(sum(originalResult) == sum(transformResult20));
        //OR22:源输出等于后续输出
        assertTrue(sum(originalResult) == sum(transformResult22));
    }

    private double sum(int[] array) {
        int sum = 0;
        for (int value : array) {
            sum += value;
        }
        return sum;
    }


    @Test
    public void testCase1() {
        Integer[] originalInput = {1, 3, 2, 6, 9};
        int[] originalResult = clip.clip_m(originalInput, 1, 5);
        applyMR_Assert(originalInput, originalResult);
    }

    @Test
    public void testCase2() {
        Integer[] originalInput = {2, 1, 4, 4, 2};
        int[] originalResult = clip.clip_m(originalInput, 1, 5);
        applyMR_Assert(originalInput, originalResult);
    }

    @Test
    public void testCase3() {
        Integer[] originalInput = {4, -2, 4, 6, 2};
        int[] originalResult = clip.clip_m(originalInput, 1, 5);
        applyMR_Assert(originalInput, originalResult);
    }

    @Test
    public void testCase4() {
        Integer[] originalInput = {9, 2, 1, 5, 3, 2};
        int[] originalResult = clip.clip_m(originalInput, 1, 5);
        applyMR_Assert(originalInput, originalResult);
    }

    @Test
    public void testCase5() {
        Integer[] originalInput = {-1, 9, 1, -3, -3};
        int[] originalResult = clip.clip_m(originalInput, 1, 5);
        applyMR_Assert(originalInput, originalResult);
    }

    @Test
    public void testCase6() {
        Integer[] originalInput = {8, 3, 2, 6, 2, 3};
        int[] originalResult = clip.clip_m(originalInput, 1, 5);
        applyMR_Assert(originalInput, originalResult);
    }

    @Test
    public void testCase7() {
        Integer[] originalInput = {1, 2, 3, 4, -5, 6};
        int[] originalResult = clip.clip_m(originalInput, 1, 5);
        applyMR_Assert(originalInput, originalResult);
    }

    @Test
    public void testCase8() {
        Integer[] originalInput = {1, 2, 4, 2, 7, 5};
        int[] originalResult = clip.clip_m(originalInput, 1, 5);
        applyMR_Assert(originalInput, originalResult);
    }

    @Test
    public void testCase9() {
        Integer[] originalInput = {1, 1, 2, 4, 1, 2};
        int[] originalResult = clip.clip_m(originalInput, 1, 5);
        applyMR_Assert(originalInput, originalResult);
    }

    @Test
    public void testCase10() {
        Integer[] originalInput = {-2, 3, 1, 4, 7};
        int[] originalResult = clip.clip_m(originalInput, 1, 5);
        applyMR_Assert(originalInput, originalResult);
    }
}

File: tests/clipTest.py
# 不需要 fixture 注入
from textual.events import Print
from mutants import runner
from src.clip import clip
from test_mutants_runner import CURRENT_MUTANT_FUNC
from MetamorphicTestGenerator1 import MetamorphicTestGenerator1
import pytest
import pytest_check as check

print("test_clip 执行了")

def applyMR_Assert(originalInput, originalResult):
    func = clip
    assert func is not None, "mutant_func 没有被注入"

    # MR1: 数组元置换（打乱顺序）
    transformInput1 = MetamorphicTestGenerator1.applyMR1(originalInput)
    transformResult1 = func(transformInput1, 1, 5)

    # MR2: 数组元素常数加法
    transformInput2 = MetamorphicTestGenerator1.applyMR2(originalInput)
    transformResult2 = func(transformInput2, 1, 5)

    # MR3_1: 加法单位元 0
    transformInput3_1 = MetamorphicTestGenerator1.applyMR3_1(originalInput)
    transformResult3_1 = func(transformInput3_1, 1, 5)

    # MR3_2: 乘法单位元 1
    transformInput3_2 = MetamorphicTestGenerator1.applyMR3_2(originalInput)
    transformResult3_2 = func(transformInput3_2, 1, 5)

    # MR5: 数组缩放变换
    transformInput5 = MetamorphicTestGenerator1.applyMR5(originalInput, 2)
    transformResult5 = func(transformInput5, 1, 5)

    # MR6: 数组反转变换
    transformInput6 = MetamorphicTestGenerator1.applyMR6(originalInput)
    transformResult6 = func(transformInput6, 1, 5)

    # MR7_1: 所有元素乘以1
    transformInput7_1 = MetamorphicTestGenerator1.applyMR7_1(originalInput)
    transformResult7_1 = func(transformInput7_1, 1, 5)

    # MR7_2: 所有元素加0
    transformInput7_2 = MetamorphicTestGenerator1.applyMR7_2(originalInput)
    transformResult7_2 = func(transformInput7_2, 1, 5)

    # MR8: 重复输入数组
    transformInput8 = MetamorphicTestGenerator1.applyMR8(originalInput)
    transformResult8 = func(transformInput8, 1, 5)

    # MR9: 复合转换一致性
    transformInput9 = MetamorphicTestGenerator1.applyMR9(originalInput, 3)
    transformResult9 = func(transformInput9, 1, 5)

    # MR10: 单调性检验
    transformInput10 = MetamorphicTestGenerator1.applyMR10(originalInput)
    transformResult10 = func(transformInput10, 1, 5)

    # MR11: 边界值替换(把最大值替换成0)
    transformInput11 = MetamorphicTestGenerator1.applyMR11(originalInput)
    transformResult11 = func(transformInput11, 1, 5)

    # MR14: 移除元素的效果（移除最大值）
    transformInput14 = MetamorphicTestGenerator1.applyMR14(originalInput)
    transformResult14 = func(transformInput14, 1, 5)

    # MR16: 重复值稳健性
    transformInput16 = MetamorphicTestGenerator1.applyMR16(originalInput)
    transformResult16 = func(transformInput16, 1, 5)

    # MR19: 插入重复元素（Java 里你使用 applyMR19）
    transformInput19 = MetamorphicTestGenerator1.applyMR19(originalInput, 2)
    transformResult19 = func(transformInput19, 1, 5)

    # MR20: 边界值灵敏度（可能由 Double 返回，需要转换为 int 列表）
    transformInput20 = MetamorphicTestGenerator1.applyMR20(originalInput)
    transformInput20_1 = [int(x) for x in transformInput20]
    transformResult20 = func(transformInput20_1, 1, 5)

    # MR22: 应用恒等变换
    transformInput22 = MetamorphicTestGenerator1.applyMR22(originalInput)
    transformResult22 = func(transformInput22, 1, 5)

    # ---------- 辅助：把结果当作列表，比较 sum ----------
    # originalResult 以及 transformResultX 都被假设为可迭代的 int 列表（或类似）
    # 为兼容性先把结果转为 list（某些 runner 可能直接返回 tuple）
    def to_list(x):
        try:
            return list(x)
        except Exception:
            # 若是单个值（不大可能），就封装为单元素列表
            return [x]

    orig_list = to_list(originalResult)
    r1 = to_list(transformResult1)
    r2 = to_list(transformResult2)
    r3_1 = to_list(transformResult3_1)
    r3_2 = to_list(transformResult3_2)
    r5 = to_list(transformResult5)
    r6 = to_list(transformResult6)
    r7_1 = to_list(transformResult7_1)
    r7_2 = to_list(transformResult7_2)
    r8 = to_list(transformResult8)
    r9 = to_list(transformResult9)
    r10 = to_list(transformResult10)
    r11 = to_list(transformResult11)
    r14 = to_list(transformResult14)
    r16 = to_list(transformResult16)
    r19 = to_list(transformResult19)
    r20 = to_list(transformResult20)
    r22 = to_list(transformResult22)


    # ---------------- Assertions ----------------
    check.equal(sum(orig_list), sum(r1), "MR1 failed (sum should be equal)")
    check.is_true(sum(orig_list) <= sum(r2), "MR2 failed (orig <= transform2)")
    check.is_true(sum(orig_list) <= sum(r3_1), "MR3_1 failed (orig <= transform3_1)")
    check.is_true(sum(orig_list) <= sum(r3_2), "MR3_2 failed (orig <= transform3_2)")
    check.is_true(sum(orig_list) <= sum(r5), "MR5 failed (orig <= transform5)")
    check.equal(sum(orig_list), sum(r6), "MR6 failed (orig == transform6)")
    check.equal(sum(orig_list), sum(r7_1), "MR7_1 failed (orig == transform7_1)")
    check.equal(sum(orig_list), sum(r7_2), "MR7_2 failed (orig == transform7_2)")
    check.is_true(sum(orig_list) <= sum(r8), "MR8 failed (orig <= transform8)")
    check.is_true(sum(orig_list) <= sum(r9), "MR9 failed (orig <= transform9)")
    check.is_true(sum(orig_list) <= sum(r10), "MR10 failed (orig <= transform10)")
    check.is_true(sum(orig_list) >= sum(r11), "MR11 failed (orig >= transform11)")
    check.is_true(sum(orig_list) >= sum(r14), "MR14 failed (orig >= transform14)")
    check.is_true(sum(orig_list) <= sum(r16), "MR16 failed (orig <= transform16)")
    check.is_true(sum(orig_list) <= sum(r19), "MR19 failed (orig <= transform19)")
    # MR20 Java 中是 assertTrue(sum(originalResult) == sum(transformResult20));
    check.equal(sum(orig_list), sum(r20), "MR20 failed (orig == transform20)")
    check.equal(sum(orig_list), sum(r22), "MR22 failed (orig == transform22)")

@pytest.mark.parametrize("originalInput", [
    [1, 3, 2, 6, 9],
    [2, 1, 4, 4, 2],
    [4, -2, 4, 6, 2],
    [9, 2, 1, 5, 3, 2],
    [-1, 9, 1, -3, -3],
    [8, 3, 2, 6, 2, 3],
    [1, 2, 3, 4, -5, 6],
    [1, 2, 4, 2, 7, 5],
    [1, 1, 2, 4, 1, 2],
    [-2, 3, 1, 4, 7],
])
def test_clip_with_func(originalInput):
    func =clip
    assert func is not None, "mutant_func 没有被注入"
    originalResult = func(originalInput, 1, 5)
    applyMR_Assert(originalInput, originalResult)
